### PR TITLE
MINOR: remove the copy constructor of LogSegment

### PR DESCRIPTION
**Pull request by johnnychhsu**
## Context
In the `LogSegment`, the copy constructor is only used in LogLoaderTest

## Solution
Remove the copy constructor and build the logic in test 

## Test
Run `./gradlew core:test --tests kafka.log.LogLoaderTest` and it succeeded

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/log/LogLoaderTest.scala [hunk 1]
```diff
@@ -352,7 +353,10 @@ class LogLoaderTest {
       // Intercept all segment read calls
       val interceptedLogSegments = new LogSegments(topicPartition) {
         override def add(segment: LogSegment): LogSegment = {
-          val wrapper = new LogSegment(segment) {
+          val idx = LazyIndex.forOffset(LogFileUtils.offsetIndexFile(wrapperLogDir, segment.baseOffset()), segment.baseOffset(), 1000)
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/log/LogLoaderTest.scala]: How about using `Mockito.spy`? for example:

```scala
          val wrapper = Mockito.spy(segment)
          Mockito.doAnswer{ in =>
            segmentsWithReads += wrapper
            segment.read(in.getArgument(0, classOf[java.lang.Long]), in.getArgument(1, classOf[java.lang.Integer]), in.getArgument(2, classOf[java.lang.Long]), in.getArgument(3, classOf[java.lang.Boolean]))
          }.when(wrapper).read(ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any())
          Mockito.doAnswer{ in =>
            recoveredSegments += wrapper
            segment.recover(in.getArgument(0, classOf[ProducerStateManager]), in.getArgument(1, classOf[Optional[LeaderEpochFileCache]]))
          }.when(wrapper).recover(ArgumentMatchers.any(), ArgumentMatchers.any())
          super.add(wrapper)
```

**Review by chia7712 — COMMENTED** [review]
@johnnychhsu thanks for improving production code. I have left a major comment. please take a look

File: core/src/test/scala/unit/kafka/log/LogLoaderTest.scala (hunk 1 above)
Comment (johnnychhsu) [reply core/src/test/scala/unit/kafka/log/LogLoaderTest.scala]: thanks for the suggestion! let me check that

Comment (johnnychhsu) [reply core/src/test/scala/unit/kafka/log/LogLoaderTest.scala]: learnt a new framework for testing, thanks for sharing this!

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
the failed tests pass on my local.
```sh
./gradlew cleanTest core:test --tests QuorumControllerTest --tests ReplicaManagerTest
```

File: core/src/test/scala/unit/kafka/log/LogLoaderTest.scala [hunk 2]
```diff
@@ -352,19 +352,15 @@ class LogLoaderTest {
...
-              super.read(startOffset, maxSize, maxPosition, minOneMessage)
-            }
-
-            override def recover(producerStateManager: ProducerStateManager,
-                                 leaderEpochCache: Optional[LeaderEpochFileCache]): Int = {
-              recoveredSegments += this
-              super.recover(producerStateManager, leaderEpochCache)
-            }
-          }
+          val wrapper = Mockito.spy(segment)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:355]: Using a `spy` mock instead of a copy constructor is not better. What is the purpose of this change? cc @chia7712

Comment (chia7712) [reply core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:355]: > What is the purpose of this change?

It seems to me there are two reason of removing the copy constructor.

1. it does not copy all fields of `LogLoader`, so it is potential bug if we use it in the production (in the future)
2. the copy constructor is used only in testing, so it should be fine to move the related behavior to testing code

> Using a spy mock instead of a copy constructor is not better. 

pardon me, could you please share the reason to me?

Comment (ijuma) [reply core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:355]: There is a lot of magic to make mocks work and even more so for spy mocks. Replacing readable/debuggable normal code for mocks is generally a bad idea. A concrete example or the problems with mocks are the mock leaks that have caused serious test instability.

As a general rule, mocks should only be used if the mock free alternative is clearly worse in terms of verbosity, maintainability, etc. I don't think it's the case here.

Comment (chia7712) [reply core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:355]: > There is a lot of magic to make mocks work and even more so for spy mocks. Replacing readable/debuggable normal code for mocks is generally a bad idea. A concrete example or the problems with mocks are the mock leaks that have caused serious test instability.
As a general rule, mocks should only be used if the mock free alternative is clearly worse in terms of verbosity, maintainability, etc. I don't think it's the case here.

Thanks for sharing. I agree that `spy` is a black box and it could have caused serious test instability.

get back to the subject. There are alternatives which can remove the incomplete copy constructor.

1. add public getters to `LogSegment` in order to reuse the constructor (https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java#L112). The side-effect is that the testing-only code is added. However, that is better than having an incomplete copy constructor.
2. reuse the constructor with temporary objects. The is the initial approach of this PR. The disadvantage is that it needs to create weird objects, but it is fine for the tests.

I prefer option 2 to replace the `spy`. WDYT?
